pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -0,0 +1,1150 @@
+	if err != nil {
+		return nil, err
+	}
+	for _, c := range containers {
+		labelledInfo := GetContainerInfoFromLabels(c.Labels)
+		pod, found := pods[labelledInfo.PodUID]
+		if !found {
+			glog.Warningf("Can not find sandbox %v for container %s", labelledInfo.PodUID, c.Name)
+			continue
+		}
+
+		converted, err := m.toRuntimeContainer(c)
+		if err != nil {
+			glog.Warningf("Convert %s container %v failed: %v", m.runtimeType, c, err)
+			continue
+		}
+
+		pod.Containers = append(pod.Containers, converted)
+	}
+
+	// Convert map to list.
+	for _, p := range pods {
+		result = append(result, p)
+	}
+
+	return result, nil
+}
+
+func (m *kubeGenericRuntimeManager) generatePodSandboxConfig(pod *api.Pod) (*runtimeApi.PodSandboxConfig, error) {
+	podFullName := kubecontainer.BuildPodFullName(pod.Name, pod.Namespace)
+	podSandboxConfig := &runtimeApi.PodSandboxConfig{
+		Name:        &podFullName,
+		Labels:      newPodLabels(pod),
+		Annotations: pod.Annotations,
+	}
+
+	dnsServers, dnsSearches, err := m.runtimeHelper.GetClusterDNS(pod)
+	if err != nil {
+		return nil, err
+	}
+	podSandboxConfig.DnsOptions = &runtimeApi.DNSOption{
+		Servers:  dnsServers,
+		Searches: dnsSearches,
+	}
+
+	hostname, _, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
+	if err != nil {
+		return nil, err
+	}
+	podSandboxConfig.Hostname = &hostname
+
+	if pod.Spec.SecurityContext != nil {
+		securityContext := pod.Spec.SecurityContext
+		podSandboxConfig.Linux = &runtimeApi.LinuxPodSandboxConfig{
+			NamespaceOptions: &runtimeApi.NamespaceOption{
+				HostNetwork: &securityContext.HostNetwork,
+				HostIpc:     &securityContext.HostIPC,
+				HostPid:     &securityContext.HostPID,
+			},
+		}
+	}
+
+	cgroupParent := ""
+	podResourceLimits := make(map[api.ResourceName]int64)
+	podResourceRequests := make(map[api.ResourceName]int64)
+	portMappings := []*runtimeApi.PortMapping{}
+	for _, c := range pod.Spec.Containers {
+		for name, limit := range c.Resources.Limits {
+			if l, ok := podResourceLimits[name]; ok {
+				podResourceLimits[name] = l + limit.MilliValue()
+			} else {
+				podResourceLimits[name] = limit.MilliValue()
+			}
+		}
+		for name, req := range c.Resources.Requests {
+			if l, ok := podResourceRequests[name]; ok {
+				podResourceRequests[name] = l + req.MilliValue()
+			} else {
+				podResourceRequests[name] = req.MilliValue()
+			}
+		}
+
+		opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, &c, "")
+		if err != nil {
+			return nil, err
+		}
+
+		for _, port := range opts.PortMappings {
+			hostPort := int32(port.HostPort)
+			containerPort := int32(port.ContainerPort)
+			protocol := toRuntimeProtocol(port.Protocol)
+			portMappings = append(portMappings, &runtimeApi.PortMapping{
+				HostIp:        &port.HostIP,
+				HostPort:      &hostPort,
+				ContainerPort: &containerPort,
+				Protocol:      &protocol,
+				Name:          &port.Name,
+			})
+		}
+
+		cgroupParent = opts.CgroupParent
+	}
+
+	if cgroupParent != "" {
```

Comment:
Should we group all Linux-specific logic to a separate function or does it not make sense at all?